### DISCUSSION TITLE
Alternative FW for usbasp or usbtinyisp to test avrdude

### DISCUSSION THREAD

**Discussion by mcuee** [avrdudes/avrdude#1045] [General]
There are many alternative usbasp firmware out there (with support for ATmega8/8a/88/88p, with extra functions like UART or HID, or totally different function than usbasp programmer).

AVR-Doper FW (stk500 based) is supposed to support usbasp out of the box. But somehow the hex file does not work on my usbasp ATmega8A clones.
https://www.obdev.at/products/vusb/avrdoper.html

Just wondering if any one here knows alternative FW for usbasp to support avr910? By right the HW seems to be compatible. FW can be an issue as V-USB CDC implementation may not work well under modern OS as it is a violation of the USB protocol (no bulk endpoint for low speed USB device). Worse case is to use a USB to TTL converter to connect to the UART port (original usbasp HW has the UART RX/TX pins connected to the 10pin ISP connector. I think nobody wants to do this since usbasp is probably better than avr910.

If you know alternative usbasp FW for either of them to simulate other programmers (like stk500v1/v2/avr910/etc), please share here. 

There are not many alternative FW for usbtinyisp (other than to support different MCU like ATtiny2313 and ATtiny 24/44/84 or ATtiny25/45/85, but usually not function improvement, only minor things like WCID). I think USBTinyISP is probably not powerful enought to emulate many other programmers. Still avr910 should be possible with the help of a USB to TTL converter.

**Comment by mcuee** [comment]
avr910:
http://ww1.microchip.com/downloads/en/appnotes/atmel-0943-in-system-programming_applicationnote_avr910.pdf

avr910 as part of STK500 (using AT90S1200 or ATTiny2313), this is for firmware upgrade of the main AT90S8535 or ATmega8535.
https://mechatronics.me.wisc.edu/labresources/schemsNmans/STK500_Schematics.pdf

usbasp schematics:
https://www.fischl.de/usbasp/

usbtinyisp (they have a AVR Studio STK500 compatible software bridge but it does not work on my Windows 11 system).
https://learn.adafruit.com/usbtinyisp/download

AVR-Doper means to be stk500 compatible (using avr-doper HW, usbasp HW and metaboard HW)
https://www.obdev.at/products/vusb/avrdoper.html

metaboard
https://metalab.at/wiki/Metaboard

**Comment by mcuee** [comment]
For AVRISP mkii, there is a clone from LUFA for the USBAVRs. 
https://www.fourwalledcubicle.com/AVRISP.php
https://github.com/abcminiuser/lufa/tree/master/Projects/AVRISP-MKII
https://code.google.com/archive/p/micropendous/wikis/AVRISP.wiki

**Reply by MCUdude** [reply]
I have an [Olimex AVRISPmkII programmer](https://www.olimex.com/Products/AVR/Programmers/AVR-ISP-MK2/open-source-hardware), and the reason this couldn't be used as a replacement for the original AVRISPmkII where I work is becase the -O flag (oscillator calibration, OSCCAL, AVR053) wasn't supported by the LUFA firmware. [It has later been added](https://github.com/abcminiuser/lufa/commit/2f70bf7ac764b7971987ceb15c1fb17b8e46915c) to LUFA, but I haven't figured how to build a working binary for the Olimex programmer.

**Reply by mcuee** [reply]
The LUFA one seems to be pretty good. And it started in 2009.
https://www.avrfreaks.net/forum/make-your-own-avrisp-mkii

**Comment by mcuee** [comment]
Google shows there is indeed an AVR910 USB programmer here using ATmega8 and 12MHz oscillator. This may well be compatible with USBASP HW.
http://www.prottoss.com/projects/AVR910.usb.prog/avr910_usb_programmer.htm

**Comment by mcuee** [comment]
There are many alternative protocol for usbasp here -- but I am not able to build the codes yet (may need to use older toolchain along with AVR Studio 4.x)
http://commonemitter.blogspot.com/2012/03/alternative-firmware-for-usbasp.html
https://github.com/cutec-chris/usbavrlab-tool/tree/master/Firmware
https://web.archive.org/web/20120125085140/http://www.ullihome.de/wiki/USBAVRLab/index (Windows host software download, linux software not available)

```
AVR programmer with either original USBasp firmware, AVRISPmkII, STK500v2 or JTAGICEmkII emulation,
JTAG/OpenOCD interface (unfortunately: still alpha),
generic USB to digital I/O interface,
digital sniffer supporting SPI, UART and I2C protocols,
most impressive in my opinion: low-speed digital oscilloscope (230 kSps).
```
The github repo also shows firmware for AVR910.

Some of the codes use V-USB CDC-ACM class for USB serial port and that will not work under Windows 8/8.1/10/11.
http://www.recursion.jp/prose/avrcdc/

But other codes use USB HID which may work.

**Reply by mcuee** [reply]
Finally figured out a way to upload the binary (encrypted hex file) to the bootloader using the Windows host app (need to program the bootloader hex file first). I tried the stk500v2 firmware and it does not work under Windows 11 as expected (due to low speed USB CDC-ACM not compliant to USB specifications). But it does work under Linux.

```
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_git -c stk500v2 -P /dev/ttyACM0 -qqp m328p 
-U ../hex/blink_uno_with_bootloader.hex && echo OK
OK

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_git -c stk500v2 -P /dev/ttyACM0 -qqp m328p 
-U flash:v:../hex/blink_uno_with_bootloader.hex && echo OK
avrdude_git: stk500v2_ReceiveMessage(): timeout
OK

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_git -c usbasp -qqp m328p 
-U flash:v:../hex/blink_uno_with_bootloader.hex && echo OK
OK
```

**Reply by mcuee** [reply]
The JTAGICE mkii FW is also recognized under Linux -- I will test it out later with a JTAG enabled device. The FW does not work under Windows (USB device not recognized).

```
mcuee@UbuntuSwift3:~$ lsusb -vvv -d 03eb:2103

Bus 003 Device 021: ID 03eb:2103 Atmel Corp. JTAG ICE mkII
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass          255 Vendor Specific Class
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         8
  idVendor           0x03eb Atmel Corp.
  idProduct          0x2103 JTAG ICE mkII
  bcdDevice            1.00
  iManufacturer           1 Atmel
  iProduct                2 JTAGICE mkII
  iSerial                 3 ULLIHOME.DEE
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0020
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower               90mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval              10
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval              10
Device Status:     0x0000
  (Bus Powered)


```

**Reply by mcuee** [reply]
The emulation of AVRISP mkii seems to be fine under Linux as well (not workng under Windows 11 -- USB device not recognized just like the other two firmware).

```
mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_git -c avrispmkii -qqp m328p
 -U flash:w:../hex/blink_uno_with_bootloader.hex && echo OK
OK

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_git -c avrispmkii -qqp m328p
 -U flash:v:../hex/blink_uno_with_bootloader.hex && echo OK
OK

mcuee@UbuntuSwift3:~/build/avr/avrdude_bin$ ./avrdude_git -c usbasp -qqp m328p
 -U flash:v:../hex/blink_uno_with_bootloader.hex && echo OK
OK


```

**Reply by mcuee** [reply]
Fixed the firmware build and build all the hex files here.
https://github.com/mcuee/usbavrlab-tool/releases/tag/snapshot_july2022

**Comment by mcuee** [comment]
usbasp stock FW will not support AT89S51 and AT89S52. Then the Chinese FW which supports AT89S51 and AT89S52 will not work with avrdude (they work with the Chinese host software like Prog ISP.

Luckily there is an alternative FW here which supports avrdude well (need to use MinGW build and not MSVC build under Windows). You need to use version 1.11 or later.
https://github.com/dioannidis/usbasp

**Comment by mcuee** [comment]
Alternative FW for USBASP to emulate AVR-Doper, from @Nibi-git.
* https://github.com/avrdudes/avrdude/issues/1626#issuecomment-1921126325

The archive contains the compiled main.hex and sources. The firmware is configured to use the HID interface.
Atmega8/8A, 12 Mhz
[firmware.zip](https://github.com/avrdudes/avrdude/files/14124635/firmware.zip)
